firestore/integration_test_internal/src/firestore_integration_test.h

```diff
@@ -43,11 +44,15 @@ bool ProcessEvents(int msec);
 // enum, but this function will gracefully handle the case where it is not.
 std::string ToFirestoreErrorCodeName(int error_code);
 
-// Waits for a Future to complete. If a timeout is reached then this method
-// returns as if successful; therefore, the caller should verify the status of
-// the given Future after this function returns. Returns the number of cycles
-// that were left before a timeout would have occurred.
-int WaitFor(const FutureBase& future);
+// Blocks until either the given predicate `pred` returns `true` or timeout
+// occurs. Returns `true` on success, `false` on timeout or if exit signal was
+// received.
+template <typename PredT>
+bool WaitUntil(const PredT& pred, int timeout_ms = kTimeOutMillis);
```

Comment:
Consider returning an enum from `WaitUntil()` and `WaitUntilFutureCompletes()` instead of conflating timeout and exit signal with a `false` return value. This will improve readability at the call sites since it is not immediately obvious without reading the function's documentation or implementation what the return value means and will avoid misleading "timeout" error messages when the exit signal occurs.

Reply:
I thought about this but was leaning towards not doing it. It seems to complicate the common case for the sake of a very rare occurrence. I'm not sure this even works correctly -- e.g. on Linux the integration tests seem to completely ignore `Ctrl-C`.

Reply:
> I'm not sure this even works correctly -- e.g. on Linux the integration tests seem to completely ignore Ctrl-C.

Please disregard -- this is actually working correctly. I'm still hesitant to complicate the common case, though.

Reply:
Hmm, it seems like it actually works in some tests and doesn't work in others.

Reply:
My concern still stands, that returning a `bool` detracts from readability at the call sites. And confounding exit signal with timeout could lead to misleading test failure messages. When tests timeout they are already hard enough to debug and unclear error messages can further make that harder. This doesn't block my approval of this PR though.

Reply:
It seems to me that the root cause here is that these functions just don't have a great way of reporting this event, which really seems more akin to an exception. There also isn't much tests can do about it, so I'm not sure they need the ability to check for this condition. I think that this type of exit usually happens due to user input, so it shouldn't be confusing to the user, at least in the general case.

Reply:
Maybe we *should* throw an exception now that exceptions are enabled in tests, especially once #561 is merged. In any case, that's out of scope for this PR so I'm fine with this code as-is.